README.md
# pipeline-editor-s5

[![Total alerts](https://img.shields.io/lgtm/alerts/g/sea-kg/pipeline-editor-s5.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/sea-kg/pipeline-editor-s5/alerts/) [![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/sea-kg/pipeline-editor-s5.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/sea-kg/pipeline-editor-s5/context:javascript) [![npm](https://img.shields.io/npm/v/pipeline-editor-s5)](https://www.npmjs.com/package/pipeline-editor-s5) [![deepcode](https://www.deepcode.ai/api/gh/badge?key=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwbGF0Zm9ybTEiOiJnaCIsIm93bmVyMSI6InNlYS1rZyIsInJlcG8xIjoicGlwZWxpbmUtZWRpdG9yLXM1IiwiYXV0aG9ySWQiOjE1NjQxLCJpYXQiOjE2Mjc3MjIxNjV9.uq45k4n60jyxIaq3CypWoApBocaB_KQRwYbbITr-54s)](https://www.deepcode.ai/app/gh/sea-kg/pipeline-editor-s5/_/dashboard?utm_content=gh%2Fsea-kg%2Fpipeline-editor-s5)

Pipeline Editor s5 (just ui-editor)

[example-editor](https://sea-kg.com/pipeline-editor-s5/)

![Alt text](/contrib/screenshot-01.jpg?raw=true "Screenshot 01")


## Integration

```html
<div style="overflow: scroll; height: 400px; min-height: 400px;">
    <canvas id="pipeline_diagram_canvas"></canvas>
</div>
<script src="./pipeline-editor-s5.js?ver=1"></script>
<script>
    window.render1 = new RenderPipelineEditor('pipeline_diagram_canvas', {
        // HERE configs
    });
    render1.set_data({
        "title": "Some",
        "description": "Some description",
        "nodes": {
            "aCaqNjY": {
                "n": "One",
                "d": "...",
                "i": {},
                "x": 1,
                "y": 0,
                "c": "#ffffff"
            },
            "kB3Cu9b": {
                "n": "Two",
                "d": "...",
                "i": {
                    "aCaqNjY": ""
                },
                "x": 2,
                "y": 2,
                "c": "#ffffff"
            }
        }
    });
</script>
```

## Integration to angular project

```
$ npm install --save pipeline-editor-s5
```

angular.json:
```
...
    "scripts": [
        ...,
        "node_modules/pipeline-editor-s5/html/pipeline-editor-s5.js"
    ]
...
```

add to html-template:
```
...
<div class="pipeline-viewer-container">
    <canvas id="pipeline_viewer"></canvas>
</div>
...
```

add to css:
```
.pipeline-viewer-container {
    display: none;
    overflow: scroll;
    max-height: 10px;
    height: 10px;
    min-height: 10px;
}
```

add to ts file:
```
...
declare var RenderPipelineEditor: any;
...

    show_data() : void {
        this.pipelineEditorInst = new RenderPipelineEditor(
          'pipeline_viewer'
        );
        this.pipelineEditorInst.set_data(this.pipelineData);
    }
```

## Documentation

### Events

* `onselectedblock(blockid)` - when user selected a block
* `onchanged()` - when changed something in scheme

### Enums

PIPELINE_EDITOR_S5_STATE_*:

* `PIPELINE_EDITOR_S5_STATE_MOVING_BLOCKS = 0;`
* `PIPELINE_EDITOR_S5_STATE_REMOVING_BLOCKS = 1;`
* `PIPELINE_EDITOR_S5_STATE_ADDING_BLOCKS = 2;`
* `PIPELINE_EDITOR_S5_STATE_ADDING_CONNECTIONS = 3;`

Line orientation:

* `PIPELINE_EDITOR_S5_LINE_ORIENT_NONE = 0;`
* `PIPELINE_EDITOR_S5_LINE_ORIENT_VERTICAL = 1;`
* `PIPELINE_EDITOR_S5_LINE_ORIENT_HORIZONTAL = 2;`
* `PIPELINE_EDITOR_S5_LINE_ORIENT_POINT = 3;`

Line angel:

* `PIPELINE_EDITOR_S5_LINE_ANGEL_END_LEFT = 0;`
* `PIPELINE_EDITOR_S5_LINE_ANGEL_END_RIGHT = 1;`
* `PIPELINE_EDITOR_S5_LINE_ANGEL_RIGHT_DOWN = 2;`
* `PIPELINE_EDITOR_S5_LINE_ANGEL_LEFT_DOWN = 3;`

### Class RenderPipelineEditor

Constructor, example:

`window.render = new RenderPipelineEditor(element_id_canvas);`

* `constructor(canvas_id, cfg)`
* `get_editor_state()` - will return from enum `PIPELINE_EDITOR_S5_STATE_*`
* `init_font_size()`
* `clone_object(obj)`
* `set_data(data)` - set new data-json, 
* `get_data()` - will return data-json
* `get_data_share()` - get_data and packed with LZString
* `set_data_share(data)` - data-json unpacked with LZString and set_data()
* `change_state_to_removing_blocks()`
* `change_state_to_moving_blocks()`
* `change_state_to_adding_blocks()`
* `change_state_to_adding_connections()`
* `canvas_onmouseover(event)`
* `canvas_onmouseout(event)`
* `canvas_onmouseup(event)`
* `canvas_onmousedown(event)`
* `find_block_id(x0, y0)`
* `canvas_onmousemove(event)`
* `update_meansures()` - Update meansures (like a width of blocks)
* `update_image_size()`
* `calcX_in_px(cell_x)`
* `calcY_in_px(cell_y)`
* `clear_canvas()`
* `draw_menu_scaling_buttons()` - Draw scaling value and buttons for it changing
* `draw_grid()`
* `draw_diagram_name()`
* `draw_blocks()`
* `call_onselectedblock` - private method for call function `onselectedblock` (only if defined by user)
* `correct_line(line)`
* `check_error(line, out_nodeid, in_nodeid)`
* `add_to_draw_connection(x0, x2, y0, y1, y2, out_nodeid, in_nodeid)`
* `auto_placement()`
* `draw_connections()`
* `debug_print_connection_info(obj)`
* `update_pipeline_diagram()`
* `do_connection_blocks()`
* `generate_new_blockid()` - will be generate uniq blockid with fixed possible collisions
* `add_block(_block_info_json)` - add new block by example json `{"n": "edit me", "d": "edit me", "i": {}, "x": t_x, "y": t_y}`
* `prepare_data_cards_one_cells()`
* `prepare_lines_out()`

### Class `RenderPipelineBlock`

* `constructor(nodeid, _conf)`
* `to_json()`
* `copy_from_json(_json)`
* `update_cell_xy(pos_x, pos_y)`
* `get_hash_cell_xy()`
* `get_cell_x()`
* `get_cell_y()`
* `set_color(val)`
* `get_color()`
* `set_name(name)`
* `get_name()`
* `set_description(description)`
* `get_description()`
* `update_block_meansures(_ctx)`
* `nodes_in_same_cells_reset()`
* `nodes_in_same_cells_add(node_id)`
* `get_paralax_in_cell()`
* `outcoming_reset()`
* `outcoming_add(nodeid)`
* `get_paralax_for_line(node_id)`
* `update_incoming_sort(pl_data_render)`
* `draw_block(_ctx, selectedBlockIdEditing)`

### Class `RenderPipelineConfig`

* `constructor()`
* `set_max_cell_xy(x,y)` - set max x and y coordinates block (in cells)
* `set_cell_size(width, height)`
* `get_cell_width()`
* `get_cell_height()`
* `get_diagram_padding_left()` - left padding for diagram
* `get_diagram_padding_top()` - top padding for diagram
* `set_card_size(width, height)`
* `get_card_width()`
* `get_card_height()`
* `get_radius_for_angels()`
* `get_scale_value()` - scaling value (1.0 == 100%)
* `set_scale_value(val)` - set new scale value (1.0 == 100%)
* `get_font_px_suffix()` - return suffix for set font in canvas like `px monospace,monospace`

### Class `RenderPipelineDrawedLinesCache`

* `constructor()`
* `add(line)`
* `has_collision(line)`
* `clear()`

### Class `RenderPipelineConnection`

* `constructor(line1, line2, line3, conf, out_nodeid, in_nodeid)`
* `draw(_ctx)`

### Class `RenderPipelineLine`

* `constructor(x0, y0, x1, y1)`
* `set_x0(val)`
* `set_x1(val)`
* `set_y0(val)`
* `set_y1(val)`
* `has_collision(line)`
* `draw_out_circle(_ctx, radius)`
* `draw_line(_ctx)`
* `draw_arrow(_ctx, radius) `
* `draw_arc(_ctx, radius, angle)`

## Other pipeline editors

Nice diagram and something else:

* https://github.com/elyra-ai/examples/tree/master/pipelines/hello_world_kubeflow_pipelines
* https://github.com/elyra-ai/pipeline-editor



Block-scheme editor:

https://draw.io


Plugin for Jenkins

https://www.jenkins.io/doc/book/blueocean/pipeline-editor/


PlantUML
* https://plantuml.com/ru/starting
* http://www.plantuml.com/plantuml/uml/SyfFKj2rKt3CoKnELR1Io4ZDoSa70000


CHANGELOG.md
# Changelog

All notable changes to pipeline-editor-s5 project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [v0.2.1] - 2021-08-07 (2021 Aug 7)


### Fixed

* Fixed #43 centering in scaling buttons
* Fixed paddings
* Fixed #39 centering name if block has not description
* Fixed #36 after delete block also delete connections


### Added

* Added hints to menu buttons
* Attached deepcode.ai

### Changed

* Redesign color for menu buttons
* Fixed #37 changed behaviour moving workspace
* Updated screenshot
* updated data_pl_sample.js
* Fixed #2 edition states complited
* Renamed render-pipeline.js to pipeline-editor-s5

### Security

* Attached deepcode.ai
* Fixed lgtm alerts

## [v0.2.0] - 2021-07-31 (2021 Jul 31)

### Fixed

* Fixed #30 forbidden moving child upper than parent
* Fixed #23 added to README integration to angular

### Changed

* Redesign format of json (minifier for share)
* Fixed #4 Scaling - redesign, moved to canvas
* Renamed RenderPipelineNode -> RenderPipelineBlock
* Start redesign modes of editor (redesign remove blocks)
* Redraw icons
* Redesign padding option for #4
* Scale value moved to RenderPipelineConfig
* Prepare #32 - reposition nodes (auto_placement)
* Updated html/images/share.svg

### Added 

* Fixed #3: autosave
* Fixed #15, #18: share by url (`?v=data_share`)
* Implemented get_data_share/set_data_share for url sharing
* added `version` to RenderPipelineEditor
* Added cursor on delete blocks
* Prepare list of functions for documentation
* Implemented settings title, description, background-color
* Added lz-string for share urls

### Security

* Fixed lgtm alerts
* Removed unused functions


## [v0.1.1] - 2021-07-13 (2021 Jul 13)

### Changed

* Redesign selection: highlight border but not fill block
* Cleanup code

### Added

* Fixed #10 added diagram name
* Fixed #31 added background-color

### Fixed

* Fixed #28 apply monospase font for json-content (For Safary browser)

### Security

* Fixed lgtm alert

## [v0.1.0] - 2021-07-06 (2021 Jul 7)

### Changed

* Data format
* Cleanup code

## [v0.0.1] - 2021-07-06 (2021 Jul 6)

### Added

* Init first version

html/index.js

window.autosave_enabled = false;

function parse_page_params() {
    var get_params = window.location.search;
    if (get_params.length > 0) {
        get_params = get_params.slice(1);
    }
    get_params = get_params.split("&");
    var result = {};
    var _regex_param = new RegExp("(.*)=([^&#]*)");
    for(var i = 0; i < get_params.length; i++){
        if(get_params[i].trim() != ""){
            var param = _regex_param.exec(get_params[i].trim());
            // console.log("results: " + JSON.stringify(p));
            if (param == null) {
                result[decodeURIComponent(get_params[i].trim().replace(/\+/g, " "))] = '';
            } else {
                result[decodeURIComponent(param[1].replace(/\+/g, " "))] = decodeURIComponent(param[2].replace(/\+/g, " "));
            }
        }
    }
    // console.log(JSON.stringify(result));
    return result;
}

function switch_ui_to_tab(_this, _callback) {
    var els = document.getElementsByClassName('pipeline-editor-tab');
    var active_id = _this.id;
    for (var i = 0; i < els.length; i++) {
        var tab_content_id = els[i].getAttribute('tab_content_id');
        if (els[i].id == active_id) {
            els[i].classList.add("active");
            document.getElementById(tab_content_id).style.display = 'block';
        } else {
            els[i].classList.remove("active");
            document.getElementById(tab_content_id).style.display = 'none';
        }
    }
    if (_callback) {
        _callback();
    }
}

function ui_render_switch_draw_grid(el) {
    if (render.is_draw_grid) {
        render.is_draw_grid = false;
        el.classList.add('pressed')
    } else {
        render.is_draw_grid = true;
        el.classList.remove('pressed')
    }
    render.update_pipeline_diagram();
}

function ui_render_update_states() {
    const states_elements = {
        "pipeline-editor-functions-btn moving-block": PIPELINE_EDITOR_S5_STATE_MOVING_BLOCKS,
        "pipeline-editor-functions-btn remove-block": PIPELINE_EDITOR_S5_STATE_REMOVING_BLOCKS,
        "pipeline-editor-functions-btn add-block": PIPELINE_EDITOR_S5_STATE_ADDING_BLOCKS,
        "pipeline-editor-functions-btn connect-blocks": PIPELINE_EDITOR_S5_STATE_ADDING_CONNECTIONS,
    };
    for (var n in states_elements) {
        var el = document.getElementsByClassName(n)[0];
        if (render.get_editor_state() == states_elements[n]) {
            el.classList.add('pressed');
        } else {
            el.classList.remove('pressed');
        }
    }
}

function ui_render_removing_blocks(el) {
    if (render.get_editor_state() != PIPELINE_EDITOR_S5_STATE_REMOVING_BLOCKS) {
        render.change_state_to_removing_blocks();
    } else {
        render.change_state_to_moving_blocks();
    }
    ui_render_update_states();
}

function ui_render_moving_blocks(el) {
    render.change_state_to_moving_blocks();
    ui_render_update_states();
}

function ui_render_adding_blocks(el) {
    render.change_state_to_adding_blocks();
    ui_render_update_states();
}

function ui_render_connecting_blocks() {
    render.change_state_to_adding_connections();
    ui_render_update_states();
}

function make_share_url(el) {
    var _url = location.protocol + "//" + location.host + location.pathname
    _url += "?v=" + render.get_data_share();
    window.open(_url, '_blank').focus();
    // var windowFeatures = "menubar=yes,location=yes,resizable=yes,scrollbars=yes,status=yes";
    // var windowObjectReference = window.open(_url, "CNN_WindowName", windowFeatures);
}

function switch_to_tab_ui_editor(active_id) {
    var _data = JSON.parse(json_content.value);
    render.set_data(_data);
    render.update_meansures();
    render.update_pipeline_diagram();
}

function switch_to_tab_json() {
    var _data = render.get_data();
    json_content.value = JSON.stringify(_data, undefined, 4);
}

function switch_to_tab_export() {
    var _data = render.get_data();
    json_content.value = JSON.stringify(_data, undefined, 4);
}

function switch_to_tab_settings() {
    var _data = render.get_data();
    json_content.value = JSON.stringify(_data, undefined, 4);

    document.getElementById("settings_background_color").value = render.backgroundColor;
    document.getElementById("settings_diagram_name").value = render.diagram_name;
    document.getElementById("settings_diagram_description").value = render.diagram_description;
}

function switch_to_tab_about() {
    var _data = render.get_data();
    json_content.value = JSON.stringify(_data, undefined, 4);
}


function save_as_image() {
    const dataUrl = pipeline_diagram_canvas.toDataURL("png");
    var win = window.open();
    win.document.write('<iframe src="' + dataUrl  + '" frameborder="0" style="border:0; top:0px; left:0px; bottom:0px; right:0px; width:100%; height:100%;" allowfullscreen> </iframe>');
}

function save_to_localstorage() {
    var _data = render.get_data();
    _data = JSON.stringify(_data, undefined, 4);
    localStorage.setItem('_data', _data);
}

function clear_localstorage() {
    localStorage.setItem('_data', '{}');
    render.set_data({})
    render.update_meansures();
    render.update_pipeline_diagram();
}

function sample_localstorage() {
    localStorage.removeItem('_data');
    render.set_data(data_pl_example);
    render.update_meansures();
    render.update_pipeline_diagram();
}

function resize_canvas() {
    // console.log(window.innerWidth);

    var canvas_cont = document.getElementById('canvas_container');

    var left_panel = 60;
    var right_panel = 120;
    var paddings = 120;
    var height_padding = 70;
    if (render.mode_viewer) {
        right_panel = 0;
        left_panel = 0;
        paddings = 20;
        height_padding = 30;
    }

    var new_width = (window.innerWidth - left_panel - right_panel - paddings) + 'px';
    canvas_cont.style['max-width'] = new_width;
    canvas_cont.style['width'] = new_width;
    

    var new_height = (window.innerHeight - height_padding) + 'px';
    canvas_cont.style['max-height'] = new_height;
    canvas_cont.style['height'] = new_height;
}

function input_onchangename() {
    var block_id = document.getElementById("prop_block_id").value;
    if (block_id) {
        render.pl_data_render[block_id].set_name(document.getElementById("prop_name").value);
        render.pl_data_render[block_id].set_description(document.getElementById("prop_description").value);
        render.pl_data_render[block_id].set_color(document.getElementById("prop_color").value);
        // render.prepare_data_render();
        render.update_meansures();
        render.update_pipeline_diagram();
    }
}

function input_onchangediagramsettings(el) {
    var _data = JSON.parse(json_content.value);
    _data["title"] = document.getElementById("settings_diagram_name").value;
    _data["description"] = document.getElementById("settings_diagram_description").value;
    _data["background-color"] = document.getElementById("settings_background_color").value;
    json_content.value = JSON.stringify(_data, undefined, 4);
    render.set_data(_data);
    render.update_meansures();
    render.update_pipeline_diagram();
}

function render_onselectedblock(block_id) {
    if (block_id) {
        document.getElementById("prop_block_id").value = block_id;
        document.getElementById("prop_name").value = render.pl_data_render[block_id].get_name();
        document.getElementById("prop_name").removeAttribute('readonly');
        document.getElementById("prop_description").value = render.pl_data_render[block_id].get_description();
        document.getElementById("prop_description").removeAttribute('readonly');
        document.getElementById("prop_color").value = render.pl_data_render[block_id].get_color();
        document.getElementById("prop_color").removeAttribute('readonly');
    } else {
        document.getElementById("prop_block_id").value = "";
        document.getElementById("prop_name").value = "";
        document.getElementById("prop_name").setAttribute('readonly', true);
        document.getElementById("prop_description").value = "";
        document.getElementById("prop_description").setAttribute('readonly', true);
        document.getElementById("prop_color").value = "";
        document.getElementById("prop_color").setAttribute('readonly', true);
    }
}

function render_onchanged() {
    if (window.autosave_enabled) {
        save_to_localstorage();
    }
}

document.addEventListener("DOMContentLoaded", function() {
    var _data = {};
    var _params = parse_page_params();
    if (_params["v"] !== undefined) {
        // view mode
        window.render = new RenderPipelineEditor('pipeline_diagram_canvas', {
            "mode-viewer": true,
        });
        render.set_data_share(_params["v"]);
        document.getElementById("container_functions").style.display = "none";
        document.getElementById("container_properties").style.display = "none";
        document.getElementById("container_editor_tabs").style.display = "none";
        document.getElementById("tab_content_ui_editor").style.top = "5px";
        document.getElementById("tab_content_ui_editor").style.height = "calc(100% - 55px)";
        render.is_draw_grid = false;
    } else {
        // editor mode
        var _data = localStorage.getItem('_data');
        if (_data) {
            _data = JSON.parse(_data);
        } else {
            _data = data_pl_example;
        }
        window.render = new RenderPipelineEditor('pipeline_diagram_canvas', {
            "mode-viewer": false,
        });
        render.set_data(_data);
    }

    resize_canvas();
    render.update_meansures();
    render.update_pipeline_diagram();
    render_onselectedblock(null);


    document.getElementById("prop_name").addEventListener('keyup', input_onchangename);
    document.getElementById("prop_description").addEventListener('keyup', input_onchangename);
    document.getElementById("prop_color").addEventListener('keyup', input_onchangename);

    var el_settings = document.getElementsByClassName("diagram-settings-input");

    for(var i = 0; i < el_settings.length; i++) {
        el_settings[i].addEventListener('keyup', input_onchangediagramsettings);
    }
    document.getElementById("settings_autosave").value = localStorage.getItem("autosave") || "off";
    window.autosave_enabled = (document.getElementById("settings_autosave").value == "on");
    document.getElementById("settings_autosave").onchange = function(el) {
        var v = this.options[this.selectedIndex].value;
        window.autosave_enabled = (v == "on");
        localStorage.setItem("autosave", v);
    };

    render.onselectedblock = render_onselectedblock;
    render.onchanged = render_onchanged;
    ui_render_update_states();
});


window.addEventListener("resize", resize_canvas);


html/data_pl_sample.js
var data_pl_example = {
    "title": "Data Example Diagram",
    "description": "data_pl_sample.js",
    "background-color": "#F6F7F1",
    "nodes": {
        "1": {
            "n": "Node",
            "d": "",
            "i": {
                "9": ""
            },
            "x": 7,
            "y": 12,
            "c": "#ffffff"
        },
        "2": {
            "n": "Node",
            "d": "",
            "i": {
                "29": ""
            },
            "x": 9,
            "y": 10,
            "c": "#ffffff"
        },
        "3": {
            "n": "Node",
            "d": "",
            "i": {
                "1": ""
            },
            "x": 6,
            "y": 14,
            "c": "#ffffff"
        },
        "4": {
            "n": "Packing",
            "d": "",
            "i": {
                "1": ""
            },
            "x": 8,
            "y": 14,
            "c": "#ffffff"
        },
        "5": {
            "n": "Node",
            "d": "",
            "i": {
                "25": ""
            },
            "x": 13,
            "y": 14,
            "c": "#ffffff"
        },
        "6": {
            "n": "Upload",
            "d": "prod",
            "i": {
                "4": ""
            },
            "x": 7,
            "y": 16,
            "c": "#ffffff"
        },
        "7": {
            "n": "Upload",
            "d": "dev",
            "i": {
                "4": ""
            },
            "x": 9,
            "y": 16,
            "c": "#ffffff"
        },
        "8": {
            "n": "Node 8",
            "d": "",
            "i": {
                "11": ""
            },
            "x": 5,
            "y": 6,
            "c": "#ffffff"
        },
        "9": {
            "n": "Node",
            "d": "",
            "i": {
                "16": ""
            },
            "x": 8,
            "y": 10,
            "c": "#ffffff"
        },
        "10": {
            "n": "Node",
            "d": "",
            "i": {
                "9": ""
            },
            "x": 8,
            "y": 12,
            "c": "#ffffff"
        },
        "11": {
            "n": "send",
            "d": "webhook",
            "i": {
                "EKeObF6": ""
            },
            "x": 6,
            "y": 4,
            "c": "#ffffff"
        },
        "15": {
            "n": "Node 15",
            "d": "",
            "i": {
                "8": "",
                "11": ""
            },
            "x": 3,
            "y": 8,
            "c": "#ffffff"
        },
        "16": {
            "n": "Node",
            "d": "",
            "i": {
                "11": "",
                "79": ""
            },
            "x": 8,
            "y": 8,
            "c": "#ffffff"
        },
        "17": {
            "n": "Node",
            "d": "",
            "i": {
                "11": ""
            },
            "x": 14,
            "y": 6,
            "c": "#ffffff"
        },
        "18": {
            "n": "Node",
            "d": "",
            "i": {
                "11": ""
            },
            "x": 15,
            "y": 6,
            "c": "#ffffff"
        },
        "19": {
            "n": "Node",
            "d": "",
            "i": {
                "11": ""
            },
            "x": 18,
            "y": 6,
            "c": "#ffffff"
        },
        "20": {
            "n": "Node",
            "d": "",
            "i": {
                "11": ""
            },
            "x": 20,
            "y": 6,
            "c": "#ffffff"
        },
        "21": {
            "n": "Node",
            "d": "",
            "i": {
                "11": ""
            },
            "x": 21,
            "y": 6,
            "c": "#ffffff"
        },
        "22": {
            "n": "Node",
            "d": "",
            "i": {
                "11": ""
            },
            "x": 22,
            "y": 6,
            "c": "#ffffff"
        },
        "23": {
            "n": "Node",
            "d": "",
            "i": {
                "25": ""
            },
            "x": 14,
            "y": 14,
            "c": "#ffffff"
        },
        "24": {
            "n": "Node",
            "d": "",
            "i": {
                "27": ""
            },
            "x": 11,
            "y": 14,
            "c": "#ffffff"
        },
        "25": {
            "n": "Node",
            "d": "",
            "i": {
                "31": ""
            },
            "x": 13,
            "y": 12,
            "c": "#ffffff"
        },
        "27": {
            "n": "Node",
            "d": "",
            "i": {
                "51": "",
                "52": "",
                "53": ""
            },
            "x": 11,
            "y": 12,
            "c": "#ffffff"
        },
        "29": {
            "n": "Node",
            "d": "",
            "i": {
                "79": ""
            },
            "x": 10,
            "y": 8,
            "c": "#ffffff"
        },
        "31": {
            "n": "Node",
            "d": "",
            "i": {
                "17": ""
            },
            "x": 14,
            "y": 9,
            "c": "#ffffff"
        },
        "32": {
            "n": "Node",
            "d": "",
            "i": {
                "31": ""
            },
            "x": 14,
            "y": 11,
            "c": "#ffffff"
        },
        "33": {
            "n": "Node",
            "d": "",
            "i": {
                "31": ""
            },
            "x": 15,
            "y": 11,
            "c": "#ffffff"
        },
        "37": {
            "n": "Node",
            "d": "",
            "i": {
                "19": ""
            },
            "x": 17,
            "y": 8,
            "c": "#ffffff"
        },
        "38": {
            "n": "Node",
            "d": "",
            "i": {
                "19": ""
            },
            "x": 19,
            "y": 8,
            "c": "#ffffff"
        },
        "40": {
            "n": "Node",
            "d": "",
            "i": {
                "37": ""
            },
            "x": 16,
            "y": 10,
            "c": "#ffffff"
        },
        "41": {
            "n": "Node",
            "d": "",
            "i": {
                "37": ""
            },
            "x": 18,
            "y": 10,
            "c": "#ffffff"
        },
        "43": {
            "n": "Node",
            "d": "",
            "i": {
                "41": ""
            },
            "x": 17,
            "y": 12,
            "c": "#ffffff"
        },
        "44": {
            "n": "Node",
            "d": "",
            "i": {
                "41": ""
            },
            "x": 19,
            "y": 12,
            "c": "#ffffff"
        },
        "45": {
            "n": "Node",
            "d": "",
            "i": {
                "59": ""
            },
            "x": 21,
            "y": 10,
            "c": "#ffffff"
        },
        "46": {
            "n": "Node",
            "d": "",
            "i": {
                "45": ""
            },
            "x": 20,
            "y": 12,
            "c": "#ffffff"
        },
        "47": {
            "n": "Node",
            "d": "",
            "i": {
                "45": ""
            },
            "x": 22,
            "y": 12,
            "c": "#ffffff"
        },
        "48": {
            "n": "Node",
            "d": "",
            "i": {
                "40": ""
            },
            "x": 16,
            "y": 13,
            "c": "#ffffff"
        },
        "49": {
            "n": "Node",
            "d": "",
            "i": {
                "48": ""
            },
            "x": 16,
            "y": 15,
            "c": "#ffffff"
        },
        "50": {
            "n": "Node",
            "d": "",
            "i": {
                "9": ""
            },
            "x": 9,
            "y": 12,
            "c": "#ffffff"
        },
        "51": {
            "n": "Node",
            "d": "",
            "i": {
                "29": ""
            },
            "x": 10,
            "y": 10,
            "c": "#ffffff"
        },
        "52": {
            "n": "Node",
            "d": "",
            "i": {
                "29": ""
            },
            "x": 11,
            "y": 10,
            "c": "#ffffff"
        },
        "53": {
            "n": "Node",
            "d": "",
            "i": {
                "29": ""
            },
            "x": 12,
            "y": 10,
            "c": "#ffffff"
        },
        "59": {
            "n": "Node",
            "d": "",
            "i": {
                "21": ""
            },
            "x": 21,
            "y": 8,
            "c": "#ffffff"
        },
        "60": {
            "n": "Node",
            "d": "",
            "i": {
                "63": ""
            },
            "x": 6,
            "y": 12,
            "c": "#ffffff"
        },
        "61": {
            "n": "Node 61",
            "d": "",
            "i": {
                "15": ""
            },
            "x": 1,
            "y": 10,
            "c": "#ffffff"
        },
        "63": {
            "n": "Node 63",
            "d": "",
            "i": {
                "15": ""
            },
            "x": 6,
            "y": 10,
            "c": "#ffffff"
        },
        "65": {
            "n": "Node 65",
            "d": "",
            "i": {
                "61": ""
            },
            "x": 0,
            "y": 12,
            "c": "#ffffff"
        },
        "66": {
            "n": "Node",
            "d": "",
            "i": {
                "61": ""
            },
            "x": 2,
            "y": 12,
            "c": "#ffffff"
        },
        "67": {
            "n": "Node",
            "d": "",
            "i": {
                "61": ""
            },
            "x": 3,
            "y": 12,
            "c": "#ffffff"
        },
        "68": {
            "n": "Node",
            "d": "",
            "i": {
                "61": ""
            },
            "x": 4,
            "y": 13,
            "c": "#ffffff"
        },
        "69": {
            "n": "Node",
            "d": "",
            "i": {
                "61": ""
            },
            "x": 5,
            "y": 13,
            "c": "#ffffff"
        },
        "71": {
            "n": "Node",
            "d": "",
            "i": {
                "66": ""
            },
            "x": 1,
            "y": 14,
            "c": "#ffffff"
        },
        "72": {
            "n": "Node",
            "d": "",
            "i": {
                "66": ""
            },
            "x": 3,
            "y": 14,
            "c": "#ffffff"
        },
        "74": {
            "n": "Node",
            "d": "",
            "i": {
                "71": ""
            },
            "x": 0,
            "y": 16,
            "c": "#ffffff"
        },
        "75": {
            "n": "Node",
            "d": "",
            "i": {
                "71": ""
            },
            "x": 2,
            "y": 16,
            "c": "#ffffff"
        },
        "79": {
            "n": "Node 79",
            "d": "",
            "i": {},
            "x": 9,
            "y": 6,
            "c": "#ffffff"
        },
        "80": {
            "n": "",
            "d": "",
            "i": {
                "EKeObF6": ""
            },
            "x": 8,
            "y": 2,
            "c": "#ffffff"
        },
        "81": {
            "n": "Node",
            "d": "",
            "i": {
                "22": ""
            },
            "x": 24,
            "y": 8,
            "c": "#ffffff"
        },
        "82": {
            "n": "Node",
            "d": "",
            "i": {
                "81": ""
            },
            "x": 24,
            "y": 10,
            "c": "#ffffff"
        },
        "83": {
            "n": "Node",
            "d": "",
            "i": {
                "81": ""
            },
            "x": 29,
            "y": 10,
            "c": "#ffffff"
        },
        "84": {
            "n": "Node",
            "d": "",
            "i": {
                "81": ""
            },
            "x": 30,
            "y": 10,
            "c": "#ffffff"
        },
        "87": {
            "n": "Node",
            "d": "",
            "i": {
                "82": ""
            },
            "x": 23,
            "y": 12,
            "c": "#ffffff"
        },
        "88": {
            "n": "Node",
            "d": "",
            "i": {},
            "x": 24,
            "y": 14,
            "c": "#ffffff"
        },
        "90": {
            "n": "some 1111",
            "d": "descrp 111",
            "i": {
                "88": "",
                "iACVkUp": ""
            },
            "x": 25,
            "y": 16,
            "c": "#ffffff"
        },
        "91": {
            "n": "Node",
            "d": "",
            "i": {
                "87": ""
            },
            "x": 23,
            "y": 14,
            "c": "#ffffff"
        },
        "92": {
            "n": "Node",
            "d": "",
            "i": {
                "iACVkUp": ""
            },
            "x": 27,
            "y": 14,
            "c": "#ffffff"
        },
        "93": {
            "n": "Node",
            "d": "",
            "i": {
                "iACVkUp": ""
            },
            "x": 28,
            "y": 14,
            "c": "#ffffff"
        },
        "94": {
            "n": "Node",
            "d": "",
            "i": {
                "iACVkUp": ""
            },
            "x": 29,
            "y": 14,
            "c": "#ffffff"
        },
        "95": {
            "n": "Node",
            "d": "",
            "i": {
                "83": ""
            },
            "x": 31,
            "y": 12,
            "c": "#ffffff"
        },
        "97": {
            "n": "Node",
            "d": "",
            "i": {
                "95": ""
            },
            "x": 30,
            "y": 14,
            "c": "#ffffff"
        },
        "98": {
            "n": "Node",
            "d": "",
            "i": {
                "95": ""
            },
            "x": 31,
            "y": 14,
            "c": "#ffffff"
        },
        "iACVkUp": {
            "n": "Node",
            "d": "",
            "i": {
                "82": ""
            },
            "x": 26,
            "y": 12,
            "c": "#ffffff"
        },
        "MXGgtjs": {
            "n": "Node",
            "d": "",
            "i": {
                "95": ""
            },
            "x": 32,
            "y": 14,
            "c": "#ffffff"
        },
        "EKeObF6": {
            "n": "Product 1",
            "d": "commit",
            "i": {},
            "x": 7,
            "y": 0,
            "c": "#4be919"
        },
        "TnGUgwz": {
            "n": "111",
            "d": "111",
            "i": {
                "80": ""
            },
            "x": 10,
            "y": 4,
            "c": "#ffffff"
        }
    }
};